ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java

```diff
@@ -5700,11 +5701,26 @@ public HiveMetaHook getHook(
       }
     };
 
-    if (conf.getBoolVar(ConfVars.METASTORE_FASTPATH)) {
-      return new SessionHiveMetaStoreClient(conf, hookLoader, allowEmbedded);
-    } else {
-      return RetryingMetaStoreClient.getProxy(conf, hookLoader, metaCallTimeMap,
-          SessionHiveMetaStoreClient.class.getName(), allowEmbedded);
+    return createMetaStoreClientFactory(conf)
+        .createMetaStoreClient(conf, hookLoader, allowEmbedded, metaCallTimeMap);
+  }
+
+  private static HiveMetaStoreClientFactory createMetaStoreClientFactory(HiveConf conf) throws
+          MetaException {
+    String metaStoreClientFactoryClassName = MetastoreConf.getVar(conf,
+        MetastoreConf.ConfVars.METASTORE_CLIENT_FACTORY_CLASS);
```

Comment:
I think instead of providing a custom factory you need to supply the custom client impl. Please take a look at https://github.com/apache/hive/pull/4257/files#diff-6561f3987ba0c11e6a8998efcdc862d3d3340d4babbe003ae8da98b1e4020faf

Reply:
Agreed! The code will be clearer if we instantiate the client impl in a single factory, maybe just named `HiveMetaStoreClientFactory `.

Reply:
I think we just need `metastore.client.class` parameter, and make a static method like `newClient(Configuration conf)` in `HiveMetaStoreClientFactory`.

Reply:
Can we simplify the match pattern as follows?
```java
String className = MetastoreConf.getVar(conf, MetastoreConf.ConfVars.METASTORE_CLIENT_CLASS);
if (conf.getBoolVar(ConfVars.METASTORE_FASTPATH)) {
  return HiveMetaStoreClientFactory.create(...);
} else {
  return RetryingMetaStoreClient.getProxy(conf, hookLoader, metaCallTimeMap, className, allowEmbedded);
}
```

Reply:
I have two thoughts.
- I think we have to explicitly disable `hive.metastore.fastpath` when configuring `metastore.client.class` since [RetryingMetaStoreClient](https://github.com/apache/hive/blob/master/standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/RetryingMetaStoreClient.java) is tightly coupled with `SessionHiveMetaStoreClient`
    - It could be acceptable if we change [the description of hive.metastore.fastpath](https://github.com/apache/hive/blob/rel/release-4.0.0-alpha-2/common/src/java/org/apache/hadoop/hive/conf/HiveConf.java#L937-L940)
- I guess people in the world already depend on `hive.metastore.client.factory.class` since we have failed to merge the patch for 7 years... This is the biggest headache